### ISSUE TITLE
Implement Sapling viewing keys

### ISSUE THREAD

**Issue by str4d** [zcash/zcash#3060]
Both incoming and full viewing keys should be storable in the wallet, supported by the RPC interface, and used to detect incoming and outgoing transactions as appropriate.

Depends on #3058, #3059, and #3061.

**Comment by str4d** [comment]
We need to circle in @lindanlee to have a discussion about the UX issues around this.

**Comment by lindanlee** [comment]
@str4d sounds good. When is appropriate? Who should I invite?

**Comment by Tomas-M** [comment]
In order for github to list this thread when searching for z_exportviewingkey, I am adding this keyword here: z_exportviewingkey

**Comment by Ehsan1605** [comment]
pls unsubscribe me

**Comment by Tomas-M** [comment]
@Ehsan1605
You can unsubscribe yourself from notifications, just go to https://github.com/zcash/zcash/issues/3060 and find "Notifications" on the right side column, there you can click unsubscribe button.

**Comment by Tomas-M** [comment]
Is there any news on implementing z_exportviewingkey RPC call for sapling?
I will be more than happy to test.

**Comment by Tomas-M** [comment]
@str4d
@daira 
It looks like all the other issues which this one depends on were closed.
Is there anything else blocking?
Thank you

**Comment by bmlis** [comment]
Heya guys, sorry to be the killjoy, but are there any rough estimates on when this functionality may arrive?

**Comment by bmlis** [comment]
For now I've managed to implement importing/exporting iv keys on my own by forking the official node. It's not unit tested yet though. I may be able to link it later, but bare in mind it's not official and only a temporary solution (I'm not sure I didn't broke anything along the way since this was my first clash with the project)

**Comment by bmlis** [comment]
As mentioned above, we prepared a minimal implementation of support for sapling ivk. I would gladly take any critique because even if that won't be merged into the upstream, we will probably end up using it until official support arrives.

**Comment by daira** [comment]
When the format for Sapling incoming viewing keys (section 5.6.6 of the protocol spec) was defined, I overlooked that there would not be sufficient information to tell which address it was meant for. We'll need to consider whether the external format should change in order to include that information, before merging #3822 or equivalent.

**Comment by jayhinz** [comment]
Thanks for responding! Any idea on how high of a priority this is for the dev team?

**Comment by zancas** [comment]
@bmlis is your implementation public?  Can you provide a link?

**Comment by ioptio** [comment]
Just a note to answer @cryptskip that ECC doesn't have a timeline/ETA at present for viewing key support.

**Comment by luboda** [comment]
But for users of this amazing tech this support is missing it could be better streamlined? Go on Thanks.

**Comment by Tomas-M** [comment]
ZCash reference wallet (light client protocol) will most likely rely on viewing keys:
https://z.cash/blog/zcash-reference-wallet-light-client-protocol/
Relying on this functionality while still not having it implemented is strange :-)

**Comment by bmlis** [comment]
Will the timeline be discussed in the near future? Encouraging moving to sapling while not providing basic API for it pushed us to creating our own fork of the node with ad-hoc functionality that is probably not very specification-compliant. Sapling is at this point a couple of months old update and node still doesn't provide full basic interface. While I understand you probably have a lot of different and more important work around zcash, it would be good to know that this will be at least worked on soon and isn't an abandoned issue.

**Comment by daira** [comment]
It is not abandoned.  I'll try to get to the issue I described in https://github.com/zcash/zcash/issues/3060#issuecomment-460367809 soon.

**Comment by bmlis** [comment]
@daira Great to hear that, thanks for the response and good luck in that case :)

**Comment by cryptskip** [comment]
Glad to hear it is not abandoned... @ioptio if ECC does end up coming up with a timeline, please share here. Thanks!

**Comment by SalineTear** [comment]
Is this open / closed / completed ?

Does Ycash have Viewing Keys and Zcash does not?

**Comment by str4d** [comment]
Most of this issue is completed. Zcash has viewing keys, zcashd uses them internally, and they are also used in the reference wallet (and ZecWallet by extension). The remaining piece for zcashd is exposing them through the RPC (which is what I believe Ycash has done on top of our existing implementation). Our implementation was blocking on resolving UX issues, but I think that's mostly resolved now? So likely just needs time to work on it.

**Comment by tromer** [comment]
There is also the matter of analyzing and explaining the security properties of the various viewing keys. Succinctly summarizing discussions from elsewhere:

* *What information* do is reveal by each kind of viewing key?
* Can an auditor trust the *integrity* of this information, i.e., all the recovered information correctly described notes that are indeed pertinent?
* Can an auditor trust the *completeness*, i.e., all pertinent notes are detected?

... and do these hold even if the sending party, the receiving party, a third party, or a collusion of several of those try to fool the auditor?

The latest discussions identified that the answer is _not_ a simple "yes" to all of the above. For example:
* To fool an auditor relying on viewing keys, colluding sender and receiver can send a transaction on-chain with dummy ciphertexts, and then send the note plaintext out-of-band (following the protocol spec, Section 4.17.1, bottom). This note will indeed become spendable by the recipient address, but won't be decryptable (hence not even detected) even by an auditor who has both parties' full viewing keys.

Is the requisite analysis and documentation work tracked somewhere?

**Comment by jackgavigan** [comment]
@tromer In the scenario you describe above, would it be possible for the auditor to detect that the input note has been spent, even if they can't see who it was sent to?

**Comment by tromer** [comment]
(Are we really going to have the discussion in *this* issue?...)

@dodger, if the note-with-dummy-ciphertext is sent from Alice to Bob,
_and_ Alice's auditor has her full viewing key (not just her incoming viewing key or her outgoing viewing key),
_and_ the note that Alice spent to fund the sending to Bob was "normal",
then _Alice's_ auditor can indeed notice that Alice made a weird transaction (which spent a note he knew belongs to Alice, but he can't decipher the outgoing notes).

Conversely: Bob's auditor won't notice anything (unless Bob is dumb when he spends the note). And Alice's auditor won't notice anything either,  if he has just her ivk, or he has just her ovk, or she was also colluding with whoever sent her the funds originally.

**Comment by nathan-at-least** [comment]
From https://github.com/zcash/zcash/issues/3060#issuecomment-550394305 :

> Is the requisite analysis and documentation work tracked somewhere?

Yes, it is now:

https://github.com/zcash/zips/issues/292

I've marked that new ticket as a dependency of deploying Viewing Key RPC support.

**Comment by tromer** [comment]
What's the interaction between this issue and _payment disclosures_ #2036?

At the very least there's the matter of education, since there's a false perception that viewing keys are suitable for proving that a *specific* payment was sent (here's a [random example](https://forum.zcashcommunity.com/t/z-addresses-usage-analysis/35497/5)); for that, payment disclosures are needed.

**Comment by tromer** [comment]
I think #4370 shouldn't close this issue (or be exposed as as non-experimental feature) before https://github.com/zcash/zips/issues/292 (_Specify security properties of Sapling viewing keys_) is addressed.

**Comment by daira** [comment]
@tromer: yes, that is what we had planned.

**Comment by nathan-at-least** [comment]
@tromer thanks for scrutinizing this on behalf of current and future Zcash users!

In [my earlier comment](https://github.com/zcash/zcash/issues/3060#issuecomment-553781938) I said:

> I've marked that new ticket as a prerequisite of deploying Viewing Key RPC support.

I literally meant in our development tracking system (Zenhub) I made an explicit requirement that this ticket cannot be closed until https://github.com/zcash/zips/issues/292 is successfully closed. Unfortunately we haven't found a way to make this association public, and the base github platform doesn't support adding such dependency requirements.

Separately from this ticket, I've made a plan with @steven-ecc to figure out how to make more of our development backlog, roadmap, and/or issue dependencies public.
